indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/autoscaler/LagBasedAutoScaler.java

```diff
@@ -200,84 +209,141 @@ private Runnable computeAndCollectLag()
+          "CurrentActiveTaskCount reached task count Min limit, skipping scale in action for dataSource[%s].",
+          dataSource
+      );
+      emitSkipMetric("Already at min task count", desiredActiveTaskCount);
+      return -1;
     }
 
-    if (withinProportion >= lagBasedAutoScalerConfig.getTriggerScaleInFractionThreshold()) {
-      // Do Scale in
-      int taskCount = currentActiveTaskCount - lagBasedAutoScalerConfig.getScaleInStep();
-      int actualTaskCountMin = Math.min(lagBasedAutoScalerConfig.getTaskCountMin(), partitionCount);
-      if (currentActiveTaskCount == actualTaskCountMin) {
-        log.debug("CurrentActiveTaskCount reached task count Min limit, skipping scale in action for dataSource[%s].",
-            dataSource
-        );
-        emitter.emit(metricBuilder
-                         .setDimension(
-                             SeekableStreamSupervisor.AUTOSCALER_SKIP_REASON_DIMENSION,
-                             "Already at min task count"
-                         )
-                         .setMetric(SeekableStreamSupervisor.AUTOSCALER_REQUIRED_TASKS_METRIC, taskCount));
-        return -1;
+    desiredActiveTaskCount = Math.min(desiredActiveTaskCount, actualTaskCountMax);
+    desiredActiveTaskCount = Math.max(desiredActiveTaskCount, actualTaskCountMin);
+    return desiredActiveTaskCount;
+  }
+
+  private void emitSkipMetric(final String reason, final int taskCount)
+  {
+    emitter.emit(metricBuilder
+                     .setDimension(SeekableStreamSupervisor.AUTOSCALER_SKIP_REASON_DIMENSION, reason)
+                     .setMetric(SeekableStreamSupervisor.AUTOSCALER_REQUIRED_TASKS_METRIC, taskCount));
+  }
+
+  /*
+    Finds the next largest factor of the topic's partition count that is ≥ `desiredTaskCount`.
+    If no such factor exists, return an empty optional.
+  */
+  private Optional<Integer> nearestPartitionCountFactor(final int desiredTaskCount)
+  {
+    int lo = 0;
+    int hi = partitionFactors.size();
+    while (lo < hi) {
+      final int mid = lo + (hi - lo) / 2;
+      if (partitionFactors.get(mid) >= desiredTaskCount) {
+        hi = mid;
       } else {
-        desiredActiveTaskCount = Math.max(taskCount, actualTaskCountMin);
+        lo = mid + 1;
       }
-      return desiredActiveTaskCount;
     }
-    return -1;
+
+    final Optional<Integer> factor = lo < partitionFactors.size()
+                                     ? Optional.of(partitionFactors.get(lo))
+                                     : Optional.empty();
+    log.debug("Given desiredTaskCount=[%d], found nearest task count=[%s]", desiredTaskCount, factor);
+    return factor;
   }
 
   public LagBasedAutoScalerConfig getAutoScalerConfig()
   {
     return lagBasedAutoScalerConfig;
   }
+
+  private static ImmutableList<Integer> factorize(final int partitionCount)
+  {
+    final ImmutableList.Builder<Integer> factors = new ImmutableList.Builder<>();
+    for (int i = 1; i <= partitionCount; ++i) {
```

Comment:
I guess you could make this list slightly smaller if desired, but clipping at the configured min/max.